cpp/oneapi/dal/backend/transfer.hpp

```diff
@@ -0,0 +1,91 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include "oneapi/dal/array.hpp"
+#include "oneapi/dal/backend/memory.hpp"
+
+namespace oneapi::dal::backend {
+
+#ifdef ONEDAL_DATA_PARALLEL
+template <typename T>
+inline std::tuple<array<T>, sycl::event> to_device(sycl::queue& q, const array<T>& ary) {
+    if (!ary.get_count()) {
+        return { ary, sycl::event{} };
+    }
+
+    if (ary.get_queue().has_value()) {
+        // TODO: Consider replacing by exception
+        ONEDAL_ASSERT(ary.get_queue().value() == q,
+                      "We do not support data tranfer between different queues yet");
```

Comment:
Why?

Reply:
Need to be replaced by context comparison, but it will complicate the logic below. It will not be enough to check alloc kind, we'll need to check weather devices are the same. Will do that.